src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  function deepClone(obj) {
+    return JSON.parse(JSON.stringify(obj));
+}
+
+let currentState = deepClone(state);
+const stateHistory = [];
+
+for (const action of actions) {
+    switch (action.type) {
+        case 'clear':
+            currentState = {};
+            break;
+        case 'addProperties':
+            currentState = {
+                ...currentState,
+                ...action.extraData,
+            };
+            break;
+        case 'removeProperties':
+            currentState = { ...currentState };
+            for (const key of action.keysToRemove) {
+                delete currentState[key];
+            }
+            break;
+        default:
+            throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    stateHistory.push(deepClone(currentState));
```

Comment:
It's better to push a copy of the object to the `stateHistory` array at the end of the loop cycle, outside of the `switch` block, after the current action is processed. This is already done correctly here.